File: examples/typescript/multi_city_search.ts
/**
 * Multi-city itinerary search (three legs on different dates).
 *
 * Mirrors examples/python/multi_city_search.py. Results come back as an
 * array of itineraries, each itinerary an array of FlightResult — one per
 * leg, in order. Run with:
 *   bun run multi_city_search.ts
 */
import {
  Airport,
  type FlightResult,
  FlightSearchFilters,
  FlightSegment,
  SeatType,
  SearchFlights,
  TripType,
} from "fli-js";

function inDays(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() + days);
  return d.toISOString().slice(0, 10);
}

function leg(from: Airport, to: Airport, dayOffset: number): FlightSegment {
  return new FlightSegment({
    departure_airport: [[[from, 0]]],
    arrival_airport: [[[to, 0]]],
    travel_date: inDays(30 + dayOffset),
  });
}

async function main(): Promise<void> {
  const filters = new FlightSearchFilters({
    trip_type: TripType.MULTI_CITY,
    passenger_info: { adults: 1, children: 0, infants_in_seat: 0, infants_on_lap: 0 },
    flight_segments: [
      leg(Airport.JFK, Airport.LHR, 0),
      leg(Airport.LHR, Airport.CDG, 4),
      leg(Airport.CDG, Airport.JFK, 7),
    ],
    seat_type: SeatType.ECONOMY,
  });

  const itineraries = (await new SearchFlights().search(filters, { topN: 3 })) as
    | FlightResult[][]
    | null;

  if (!itineraries?.length) {
    console.log("No itineraries found.");
    return;
  }

  itineraries.forEach((legs, i) => {
    const total = legs.reduce((sum, l) => sum + (l.price ?? 0), 0);
    console.log(`Itinerary ${i + 1} — total $${total.toFixed(0)}`);
    for (const segment of legs) {
      const first = segment.legs[0];
      const last = segment.legs[segment.legs.length - 1];
      console.log(
        `  ${first.departure_airport} -> ${last.arrival_airport}  ` +
          `$${segment.price ?? "N/A"}  ${segment.duration} min  ${segment.stops} stop(s)`,
      );
    }
  });
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/multi_city_search.py
#!/usr/bin/env python3
"""Multi-city flight search example.

This example demonstrates how to search a multi-city itinerary (three or
more one-way legs on different dates) in a single request. Multi-city
results come back as tuples of ``FlightResult`` — one entry per leg, in
the order the legs were supplied.
"""

from datetime import datetime, timedelta

from fli.models import (
    Airport,
    FlightSearchFilters,
    FlightSegment,
    PassengerInfo,
    SeatType,
    TripType,
)
from fli.search import SearchFlights


def main():
    base = datetime.now() + timedelta(days=30)

    def leg(origin: Airport, destination: Airport, day_offset: int) -> FlightSegment:
        return FlightSegment(
            departure_airport=[[origin, 0]],
            arrival_airport=[[destination, 0]],
            travel_date=(base + timedelta(days=day_offset)).strftime("%Y-%m-%d"),
        )

    # JFK -> LHR, a few days in London, LHR -> CDG, then CDG -> JFK home.
    filters = FlightSearchFilters(
        trip_type=TripType.MULTI_CITY,
        passenger_info=PassengerInfo(adults=1),
        flight_segments=[
            leg(Airport.JFK, Airport.LHR, 0),
            leg(Airport.LHR, Airport.CDG, 4),
            leg(Airport.CDG, Airport.JFK, 7),
        ],
        seat_type=SeatType.ECONOMY,
    )

    search = SearchFlights()
    # top_n controls how many outbound options each leg is expanded against.
    itineraries = search.search(filters, top_n=3)

    if not itineraries:
        print("No itineraries found.")
        return

    print(f"Found {len(itineraries)} multi-city itineraries:\n")
    for i, legs in enumerate(itineraries, 1):
        total = sum(segment.price for segment in legs)
        print(f"--- Itinerary {i} — total ${total:.0f} ---")
        for segment in legs:
            first, last = segment.legs[0], segment.legs[-1]
            print(
                f"  {first.departure_airport.value} -> {last.arrival_airport.value}"
                f"  ${segment.price:.0f}  {segment.duration} min  {segment.stops} stop(s)"
            )
        print()


if __name__ == "__main__":
    main()
